### PR TITLE
fix prepare context redundant code problem, optimize executor by cach…

### PR DESCRIPTION
**Pull request by guru4elephant**
…ing create_varaiables

test=develop

optimize executor by caching create_variable results. User only need to open use_program_cache to use this optimization. 

Take Bert model as example:

Before optimization:
image

After optimization：
image

image

This optimization is generalized that all models training can be applied.

### CODE REVIEW

File: python/paddle/fluid/executor.py [hunk 1]
```diff
@@ -719,7 +740,8 @@ def _run(self, program, exe, feed, fetch_list, feed_var_name,
         if not use_program_cache:
             exe.run(program.desc, scope, 0, True, True, fetch_var_name)
         else:
-            exe.run_cached_prepared_ctx(ctx, scope, True, True, False)
+            #exe.run_cached_prepared_ctx(ctx, scope, True, True, False)
```
Comment (gavin1332) [inline python/paddle/fluid/executor.py]: remove this line

**Review by gavin1332 — CHANGES_REQUESTED** [review]
a commented line need to be removed

File: python/paddle/fluid/executor.py (hunk 1 above)
Comment (guru4elephant) [reply python/paddle/fluid/executor.py]: fixed, thx

**Review by gavin1332 — DISMISSED** [review]

**Review by gavin1332 — DISMISSED** [review]

File: python/paddle/fluid/executor.py [hunk 2]
```diff
@@ -366,6 +366,12 @@ def __init__(self, place):
         self._default_executor = core.Executor(p)
         self._closed = False
 
+    def _get_var_cache(self, program_cache_key):
+        return self.var_caches.get(program_cache_key, None)
```
Comment (gavin1332) [inline python/paddle/fluid/executor.py:372]: missing definition of self.var_caches

File: python/paddle/fluid/executor.py [hunk 3]
```diff
@@ -366,6 +366,12 @@ def __init__(self, place):
...
+
+    def _get_scope_cache(self, program_cache_key):
+        return self.scope_caches.get(program_cache_key, None)
```
Comment (gavin1332) [inline python/paddle/fluid/executor.py:375]: missing definition of self.scope_caches

File: python/paddle/fluid/executor.py [hunk 4]
```diff
@@ -719,7 +744,7 @@ def _run(self, program, exe, feed, fetch_list, feed_var_name,
         if not use_program_cache:
             exe.run(program.desc, scope, 0, True, True, fetch_var_name)
         else:
-            exe.run_cached_prepared_ctx(ctx, scope, True, True, False)
+            exe.run_cached_prepared_ctx(ctx, scope, False, False, False)
```
Comment (gavin1332) [inline python/paddle/fluid/executor.py]: as executor.run() method allows user inputting None scope, it should be set scope to default when it is None

File: python/paddle/fluid/executor.py (hunk 2 above)
Comment (guru4elephant) [reply python/paddle/fluid/executor.py:372]: fixed. thx

File: python/paddle/fluid/executor.py (hunk 3 above)
Comment (guru4elephant) [reply python/paddle/fluid/executor.py:375]: fixed.

File: paddle/fluid/pybind/pybind.cc [hunk 5]
```diff
@@ -943,6 +943,8 @@ All parameter, weight, gradient are variables in Paddle.
            })
       .def("prepare_ctx_cache", &Executor::PrepareCtxCache,
            py::call_guard<py::gil_scoped_release>())
+      .def("create_variables", &Executor::CreateVariables,
```
Comment (jacquesqiao) [inline paddle/fluid/pybind/pybind.cc:946]: CreateVariables has no return value.

**Review by gavin1332 — APPROVED** [review]

File: paddle/fluid/pybind/pybind.cc (hunk 5 above)
Comment (guru4elephant) [reply paddle/fluid/pybind/pybind.cc:946]: will remove this later, thx

**Pull request merged**